metricbeat/module/kubernetes/state_resourcequota/state_resourcequota.go

```diff
@@ -0,0 +1,101 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package state_resourcequota
+
+import (
+	"github.com/elastic/beats/libbeat/common"
+	p "github.com/elastic/beats/metricbeat/helper/prometheus"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("kubernetes", "state_resourcequota",
+		NewResourceQuotaMetricSet,
+		mb.WithHostParser(p.HostParser))
+}
+
+// ResourceQuotaMetricSet is a prometheus based MetricSet that looks for
+// mb.ModuleDataKey prefixed fields and puts then at the module level
+//
+// Copying the code from other kube state metrics, this should be improved to
+// avoid all these ugly tricks
+type ResourceQuotaMetricSet struct {
+	mb.BaseMetricSet
+	prometheus p.Prometheus
+	mapping    *p.MetricsMapping
+}
+
+// NewResourceQuotaMetricSet returns a prometheus based metricset for ResourceQuotas
+func NewResourceQuotaMetricSet(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	prometheus, err := p.NewPrometheusClient(base)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResourceQuotaMetricSet{
+		BaseMetricSet: base,
+		prometheus:    prometheus,
+		mapping: &p.MetricsMapping{
+			Metrics: map[string]p.MetricMap{
+				"kube_resourcequota_created": p.Metric("created.sec"),
+				"kube_resourcequota":         p.Metric("quota"),
+			},
+			Labels: map[string]p.LabelMap{
+				"namespace":     p.KeyLabel(mb.ModuleDataKey + ".namespace"),
+				"resourcequota": p.KeyLabel("name"),
+
+				"resource": p.KeyLabel("resource"),
+				"type":     p.KeyLabel("type"),
+			},
+		},
+	}, nil
+}
+
+// Fetch prometheus metrics and treats those prefixed by mb.ModuleDataKey as
+// module rooted fields at the event that gets reported
+func (m *ResourceQuotaMetricSet) Fetch(reporter mb.ReporterV2) {
+	events, err := m.prometheus.GetProcessedMetrics(m.mapping)
+	if err != nil {
+		m.Logger().Error(err)
+		reporter.Error(err)
+		return
+	}
+
+	for _, event := range events {
+		var moduleFieldsMapStr common.MapStr
+		moduleFields, ok := event[mb.ModuleDataKey]
+		if ok {
+			moduleFieldsMapStr, ok = moduleFields.(common.MapStr)
+			if !ok {
+				m.Logger().Errorf("error trying to convert '%s' from event to common.MapStr", mb.ModuleDataKey)
+			}
+		}
+		delete(event, mb.ModuleDataKey)
```

Comment:
Consider using `mb.TransformMapStrToEvent()` as [here](https://github.com/elastic/beats/pull/13544/files#diff-96984df8acbe3fbcfd715d738621fd8aR112) to handle these special fields.

Reply:
this is boilerplate to me, glad to copy it from wherever I should
Thanks